cmdstanpy/model.py

```diff
@@ -1231,132 +1225,115 @@ def variational(
-        )
-        pattern_compiled = re.compile(pattern, flags=re.IGNORECASE)
-        previous_count = 0
-        stdout = b''
-        changed_description = False  # Changed from 'warmup' to 'sample'
-        pbar.set_description(desc=f'Chain {idx + 1} - warmup', refresh=True)
-
+    def _wrap_sampler_progress_hook(
+        self, chain_id: int, total: int
+    ) -> Optional[Callable[[str], None]]:
+        """Sets up tqdm callback for CmdStan sampler console msgs."""
         try:
-            # iterate while process is sampling
-            while proc.poll() is None and proc.stdout is not None:
-                output = proc.stdout.readline()
-                stdout += output
-                output = output.decode('utf-8').strip()
-                if output.startswith('Iteration'):
-                    match = re.search(pattern_compiled, output)
-                    if match:
-                        current_count = int(match.group(1))
-                        total_count = int(match.group(2))
-
-                        if pbar.total != total_count:
-                            pbar.reset(total=total_count)
-
-                        if (
-                            match.group(3).lower() == 'sampling'
-                            and not changed_description
-                        ):
-                            pbar.set_description(f'Chain {idx + 1} - sample')
-                            changed_description = True
-
-                        pbar.update(current_count - previous_count)
-                        previous_count = current_count
-
-            pbar.set_description(f'Chain {idx + 1} -   done', refresh=True)
-
-            if 'notebook' in type(pbar).__name__:
-                # In Jupyter make the bar green by closing it
-                pbar.close()
-
-        except Exception as e:  # pylint: disable=broad-except
-            get_logger().warning(
-                'Chain %s: Failed to read the progress on the fly. Error: %s',
-                idx,
-                repr(e),
+            from tqdm.autonotebook import tqdm  # type:ignore
+
+            pbar: Any = tqdm(
+                total=total,
+                bar_format="{desc} |{bar}| {postfix[0][value]}",
+                postfix=[dict(value="Status")],
+                desc=f'chain {chain_id}',
+                colour='yellow',
+                leave=False,  # cleans terminal
             )
 
-        return stdout
+            def sampler_progress_hook(line: str) -> None:
+                if line.startswith("Iteration"):
+                    if 'Sampling' in line:
+                        pbar.colour = 'blue'
+                    pbar.update(1)
+                    pbar.postfix[0]["value"] = line
+                elif line.startswith("Elapsed"):
+                    pbar.close()
+
+        except (ImportError, ModuleNotFoundError):
+            print("tqdm was not downloaded, progressbar not shown")
```

Comment:
agreed - will try again.

Reply:
the progress bar callback hook is much more responsive using a reporter class, FWIW.
I've tweaked/hacked this so that:

a) the progress bar remains after sampling has completed
b) in the terminal window, messages written to stdout expose intermediate versions of the progress bar (at least they do on my machine).  to hack this, I write out chains worth of padded blank lines, which leaves space between commands, but also, leaves things looking OK everywhere.

also, can't get chains to display in oder - what's there is as good as I can get it.

Reply:
This seems much much better for me now in testing. Order was a stretch, especially with doing everything threaded.  Thanks!